State classes

nhprof opened the pull request:
Initial draft of the conversion to state classes.  This is not intended to change the behavior in any way, only the organization of the code.  Since a lot of lines were moved around, I would appreciate scrutiny as it is quite possible that something got changed or missed.  As of the submission of this pull request, the branch compiles cleanly but has not yet been tested on the hardware.

Basic idea:  every state has a class with two methods.  
* The enter() method does everything necessary for the transition.  
* The update() method performs maintenance and checks for possible transition to the next state.

I also moved most of the global variables to a BreathData class, along with the functions that update them.

nhprof committed:
Initial file import

Still need to update contents.

nhprof committed:
Clean compile

Still untested on hardware

nhprof commented:
Anybody know what those build errors mean?  It compiles fine on my laptop.

nhprof committed:
Removed testAlarmManager folder

faridasabry2 commented:
I tried to fix it but the full Nextion library isn't available through the arduino-cli. Three possible solutions I think of: we can either investigate updating the arduino-cli somehow (submit a PR or sth maybe?),  upload all the needed Nextion library files, or remove the build workflow altogether.

phalpern commented on BreathData.h:
```diff
@@ -0,0 +1,55 @@
+    float tidalVolumeInsp = 0.0; // measured inspiratory tidal volume
+    float tidalVolumeExp = 0.0;  // measured expiratory tidal volume
+
+    float lastPeep = 0.0/0.0; // PEEP from last breath
+    float lastPeak = 0.0/0.0; // peak pressure from last breath
```
What?  Division by zero?  If you want a NaN, a cleaner way to get one is to call `std::nan("")` from `<cmath>`.  If the Arduino library doesn't have that, then you can put this into `constants.h`:
```
static const float NaNf = 0.0f/0.0f;
```
Then use the `NaNf` value to initialize these members.

nhprof replied:
This is copied verbatim from the previous version.  It looks like the NaN is used here to avoid an alarm on the first breath cycle.  Actually lastPeep is never used at all.

I've taken your NaNf suggestion.

phalpern commented on BreathData.cpp:
```diff
@@ -0,0 +1,146 @@
+#include "ProportionalValve.h"
+#include "BreathData.h"
+
+static unsigned long BreathData::cycleCount = 0; // number of breaths (including current breath)
```
The `static` keyword does not belong.  I'm surprised this compiles.  Why is this static, anyway?  Why is it not a normal member of `BreathData`?

phalpern commented on BreathData.h:
```diff
@@ -0,0 +1,55 @@
+
+    static unsigned long cycleCount; // number of breaths (including current breath)
+
+    // update functions
```
The combination of all-public data combined with a raft of update functions is suspicious.  Does this class *do* anything, or is it just a collection of related variables.  I vote for the latter, but if it actually does something, then the data should be private in order to maintain abstraction.

nhprof replied:
Agreed.

phalpern commented on BreathData.cpp:
```diff
@@ -0,0 +1,146 @@
+  // calculate actual minute volume from last breath
+  const float minuteVolume = inspFlowReader.getVolume() * CC_PER_MS_TO_LPM / cycleDuration;
+
+  // Update patient data on display to reflect values from last breath
+  display.writePeak(inspPressureReader.peak());        // measured pip cmH2O
+  display.writePlateau(inspPressureReader.plateau());  // measured plateau (only measured if HOLD_INSP_STATE is activated) cmH2O
+  display.writePeep(expPressureReader.peep());         // measured PEEP cmH2O
+  display.writeVolumeExp(expFlowReader.getVolume());   // measured expired volume
+  display.writeMinuteVolume(minuteVolume);             // measured minute volume
+  display.writeBPM(60000.0/cycleDuration);             // measured respiratory rate (seconds)
+  display.writeO2(O2);                                 // measured FIO2 concentration (@FutureWork: pending addition of O2 sensor)  
+
+  // close expiratory valve
+  expValve.close();
```
This seems like a strange combination of things to do in this class.  You are collecting data, updating the display, and moving valves all in a single function whose purpose is no longer clear just from its name.  I think these operations need to be split out so that display updates happen in a display module (that might read this data) and valve changes should be in the state machine.

nhprof replied:
These are good ideas!  I don't want to make them in this update, which was intended to focus on introducing classes to the state dynamics, and has already gotten unwieldy.  They should be addressed in the future, so maybe this should be added to the list of outstanding issues.

phalpern commented on BreathData.cpp:
```diff
@@ -0,0 +1,146 @@
+
+  // begin PID control based on desired flow and reset tidal volume
+  inspValve.beginBreath(desiredInspFlow); 
+  inspFlowReader.resetVolume();           
```
These things look like they belong to the state machine, not to the data class.

nhprof replied:
Agreed.

phalpern commented on BreathData.cpp:
```diff
@@ -0,0 +1,146 @@
+ * @param highAlarmCode -- the high alarm code for a specific sensor
+ * @param lowAlarmCode -- the low alarm code for a specific sensor
+ */ 
+void BreathData::checkAlarmRangeWithUpdate(float reading, float &compareValue, float sensitivity, alarmCode highAlarmCode, alarmCode lowAlarmCode) { 
```
Wouldn't this make more sense in the alarms module?

I'm a little uneasy about this function updating the `compareValue` this way.  An alternative way to write it, which would also avoid the need for the little function below, would be:
```
float BreathData::checkAlarmRange(float reading, float compareValue,
                                  float sensitivity, alarmCode highAlarmCode, alarmCode lowAlarmCode)
```
Which would return `compareValue` unchanged if an alarm is activated, otherwise it would return `reading`.  You would call it like this:
```
xyzLastReading = checkAlarmRange(xyzReading, xyzLastReading, ...);
```

nhprof replied:
I like this suggestion!

phalpern commented on BreathData.cpp:
```diff
@@ -0,0 +1,146 @@
+/** 
+ * check for errors in sensor readings and take appropriate action
+ */ 
+void BreathData::checkSensorReadings() {
```
Unclear why this is in this class instead of being in the sensors module.

nhprof replied:
We don't really have a sensors module yet.  This is here at the moment because it checks whether the current breath values are in alarm ranges.  Probably the name of the function should be changed to checkForAlarmConditions.

phalpern commented on State.h:
```diff
@@ -0,0 +1,82 @@
+ */
+class State {
+  public:
+    virtual State* update() = 0;           // perform maintenance, and perhaps transition to new state
```
Why isn't the `enter()` method not also a virtual function in the base class?

nhprof replied:
Now it is.

phalpern commented on State.h:
```diff
@@ -0,0 +1,82 @@
+// Standby mode.  This is the initial state.
+class OffState : public State {
+  public:
+    static State* enter();               // performs actions necessary upon entering the state
```
By making `enter` static, you are preventing it from directly setting member variables in this class.  It feels like a weird asymmetry that `enter` belongs to the class but `update` belongs to the state.  Since every state is a singleton, they can easily both be virtual functions.

nhprof replied:
I wanted `enter` to be static so you can call it from just the class name.  Otherwise you have to have all the singletons be globally available, right?   That's not a problem with `update` because you should already have a local pointer to it before updating.

phalpern replied:
Sure, but there are several ways to address that:

1. Make all the singletons global.  This is not as bad as it sounds.  The state machine needs to know all of the state names anyway, and each state name is represented by a singleton, so there isn't really any loss of abstraction nor any proliferation of names that isn't already there.
2. Have a `singleton()` static function in in each class.  This is slightly cleaner, but also a bit more cumbersome, as you'd have to write `OffState::singleton().enter()` instead of just `OffState::enter()` or `offState.enter()`.
3. An interesting variation of #1 is to create a global function template, `enterState<StateClass>()`, so you would write `enterState<OffState>()` instead of `OffState::enter()`.  `enterState` would be written as:

```
template <class StateClass>
State* enterState() {
    return State::singleton().enter();
};
```
As you can see, the asymmetry really bugs me.  I think for example that the state class should hold onto all of its own state-related timers, rather than putting all of that junk into `BreathData`.  That will work best if `enter` and `update` are both virtual functions.  Note that the individual state classes can also have non-virtual functions to return, e.g., a timer value when the state was last entered or exited.

nhprof replied:
Asymmetry resolved.

phalpern commented on State.h:
```diff
@@ -0,0 +1,82 @@
+  public:
+    static State* enter();               // performs actions necessary upon entering the state
+    State* update();                     // perform maintenance, and perhaps transition to next state
+    static State* begin(BreathData);     // returns the OffState instance with breath data link
```
The name `begin` is confusing in two ways: 1) What is it beginning? and 2) it conflicts with the standard library's use of `begin` which returns an iterator pointing to the first element of a container.  How about `instance` or `singleton` instead?

nhprof replied:
Good point; I forgot about the name conflict.  This function is supposed to be called to fire up the state machine and retrieve the first state.  I'll rename it `initState`.  Although maybe we don't even need it if I initialize breath somewhere else.  Will consider further.

phalpern commented on State.h:
```diff
@@ -0,0 +1,82 @@
+    static State* begin(BreathData);     // returns the OffState instance with breath data link
+  private: 
+    OffState() {};                       // to create singleton instance
+    static OffState *instance;           // storage for singleton instance
```
This should be unnecessary.  See my comment for `OffState::begin` in `State.cpp`.

phalpern replied:
At the very least, it should be an object, not a pointer, as I described in State.cpp.

nhprof replied:
Fixed.

phalpern commented on State.cpp:
```diff
@@ -0,0 +1,248 @@
+
+// performs actions necessary upon entering a state
+static State* OffState::begin(BreathData b) {
+    if (instance==0) instance = new OffState();  // one-time setup
```
There is no need to use pointers and allocate from the heap, nor do you need an `if` statement. Simply write:
```
  OffState instance;
...
  return &instance;
```
The compiler will do the rest more efficiently than you can.

nhprof replied:
I copied this from examples of the singleton design pattern in C++.  If we write it the way you suggest, won't it create a new instance each time we call `enter` or `begin`?

phalpern commented on State.h:
```diff
@@ -0,0 +1,82 @@
+};
+
+// Inspiration begins a breath cycle
+class InsStateVC : public State {
```
All of my comments about `OffState` apply to this state and the rest.

phalpern commented on State.cpp:
```diff
@@ -0,0 +1,248 @@
+#include "AlarmManager.h"
+#include "State.h"
+
+// Initialize singleton instances
+OffState* OffState::instance = 0;
+InsStateVC* InsStateVC::instance = 0;
+InsHoldStateVC* InsHoldStateVC::instance = 0;
+ExpStateVC* ExpStateVC::instance = 0;
+PeepStateVC* PeepStateVC::instance = 0;
+ExpHoldStateVC* ExpHoldStateVC::instance = 0;
```
All of these should go away, as per my comment on `OffState::begin`

phalpern replied:
They're still here.  Singletons don't need global pointers; they need global objects:
```
ExpStateVC ExpStateVC::instance;
```
Or better yet, make it a singleton static member function:
```
ExpStateVC& ExpStateVC::singleton()  // Static
{
    ExpStateVC theSingleton;
    return theSingleton;
}
```

phalpern commented on State.cpp:
```diff
@@ -0,0 +1,248 @@
+PeepStateVC* PeepStateVC::instance = 0;
+ExpHoldStateVC* ExpHoldStateVC::instance = 0;
+
+BreathData* State::breath;
```
Instead of being a pointer, why not just have a `BreathData` object?

nhprof replied:
I was wanting to have the BreathData object created elsewhere and passed in here, and I couldn't do that with a plain object.  I think this is another example of me trying to use Java idioms that aren't appropriate.  I'll take your advice.

phalpern commented on State.cpp:
```diff
@@ -0,0 +1,248 @@
+// performs actions necessary upon entering a state
+static State* OffState::begin(BreathData b) {
+    if (instance==0) instance = new OffState();  // one-time setup
+    breath = &b;                                 // store reference to breath data
```
This won't work. You are storing the address of a local variable, which will go out of scope immediately on return.  The `BreathData` argument needs to be a pointer.  But really, it need not exist, since the base class already has a `breath` object, there is no need to set it in every state.

nhprof replied:
It's been a long while since I've programmed in C++, so my Java-based thinking is getting me in trouble.  I just want to initialize the base-class breath object for the first time.  Not sure where to put that.

phalpern commented on State.cpp:
```diff
@@ -0,0 +1,248 @@
+    breath = &b;                                 // store reference to breath data
+
+    // valve settings
+    inspValve.endBreath(); // close the inspiratory valve
```
This is confusing. Why are there actions in something that supposedly just returns the address of a singleton?  Don't mix up accessors with action functions.

nhprof replied:
There aren't really any pure accessors here.  The idea is for the program to keep just one pointer to the current state.  The only way to get a pointer to another state is to enter it by calling `enter`.

I suppose one could take this idea further and make the state instances completely unavailable to the main program, which would then only call static methods in State.  Not sure that's an improvement, though.

phalpern commented on State.cpp:
```diff
@@ -0,0 +1,248 @@
+    return instance;
+}
+
+static State* OffState::enter() {
```
OK, at this point, I think you are mostly copying from the old state machine.  I have to go now, but this looks like it's on the right track.

phalpern reviewed (commented):
I reviewed the basic structure, but not the state-machine logic itself.  I think the main issues are that `BreathData` does too much and that the state class is using `BreathData` oddly.

nhprof committed:
Updates noted in Pablo's review

Changes to state classes, etc.
Still need reorganization/split of BreathData.

nhprof committed:
Rename checkSensorReadings

Changed to checkForAlarmConditions.

nhprof committed:
Improvement to function checking alarm conditions.

phalpern commented on State.cpp:
```diff
@@ -17,27 +21,17 @@ ExpStateVC* ExpStateVC::instance = 0;
-    // if we enter this state other than at the beginning, sound an alarm        
-    alarmMgr.activateAlarm(ALARM_SHUTDOWN);
+    if (instance==0) {
+        instance = new OffState();  // one-time setup
```
If you're calling `new`, you're doing something wrong.  This is a classic singleton. There is no reason its space cannot be allocated statically before the program even begins running.  See my comments about globals and singletons, above

Let the compiler do the work.  There is nowhere in this entire project where heap allocation is needed (except maybe for strings in the display code, but even then, I doubt it).  Note that I still believe that `enter()` should be a virtual function (or at least a non-static class function).

phalpern commented on circuit-control.ino:
```diff
@@ -162,248 +83,40 @@ void setup() {
-    } break;
-  } // End switch
+  // handle state maintenance & update
+  state = state->update();
```
Instead of having each state call `enter` for the next state, consider doing it here, in the main loop or, better yet, in a subroutine:
```
void runStateMachine() {
  State* nextState = state->update();
  while (nextState != state) {
    state = nextState;
    nextState = state->enter();
  }
}
```
Note that this neatly solves the issue of whether `enter` should be virtual.  Inside the individual states, instead of writing:
```
    nextState = InsHoldStateVC::enter();
    ...
    return nextState;
```
You would write simply:
```
    return &InsHoldStateVC::singleton();
```
I like centralizing the state machine logic this way, and I think it will be easier to maintain. Also, you can eventually make the state machine itself into a class that holds on to the current state.

nhprof replied:
As you can see, I adopted a lot of these ideas in a recent commit.

phalpern reviewed (changes requested):
It's coming along, but I think the state machine logic can be cleaner and more centralized. Don't use static pointers when static objects will do.

nhprof committed:
Further class refactoring

Created StateMachine class to run states.  Individual states become nested classes.  States no longer call enter() within their update methods; instead, this is centralized in an update() method of StateMachine.

nhprof commented:
I think the new commit incorporates most of Pablo's suggestions.  Thanks for pushing on them!  (And apologies for how long they took me to address.)  

The main outstanding issue I can see remaining is the BreathData class, which received a lot of the actual data update methods from the old code and isn't properly separated.  Although it needs to be done at some point, I don't want to address it in this pull request for two reasons:
(1) my goal was to improve the class design of the state machine, which is more or less done now
(2) a deeper modification of the data updates will likely introduce bugs, and we're not currently set up to run lab tests with the hardware so that these could be identified and fixed

nhprof committed:
Removed redundant files

Forgot to remove State.h and State.cpp, which had their contents moved to StateMachine.h and StateMachine.cpp respectively.

nhprof committed:
Fixed pointer usage

Singleton pointers weren't being returned correctly.  Switched to using a public field; now seems to work.
